src/libPersistence/ContractStorage2.cpp

```diff
@@ -140,11 +142,31 @@ bool ContractStorage2::FetchStateValue(const dev::h160& addr, const bytes& src,
     LOG_GENERAL(INFO, "query for fetch: " << query.DebugString());
   }
 
-  if (query.name() == FIELDS_MAP_DEPTH_INDICATOR) {
-    LOG_GENERAL(WARNING, "query name is " << FIELDS_MAP_DEPTH_INDICATOR);
+  if (query.name() == CONTRACT_ADDR_INDICATOR ||
+      query.name() == SCILLA_VERSION_INDICATOR ||
+      query.name() == MAP_DEPTH_INDICATOR || query.name() == TYPE_INDICATOR ||
+      query.name() == HAS_MAP_INDICATOR) {
+    LOG_GENERAL(WARNING, "invalid query: " << query.name());
     return false;
   }
 
+  if (getType) {
+    std::map<std::string, bytes> t_type;
+    std::string type_key =
+        GenerateStorageKey(addr, TYPE_INDICATOR, {query.name()});
+    FetchStateDataForKey(t_type, type_key, true);
+    if (t_type.empty()) {
+      LOG_GENERAL(WARNING, "Failed to fetch type for addr: "
+                               << addr.hex() << " vname: " << query.name());
+      return false;
+    }
+    try {
+      type = DataConversion::CharArrayToString(t_type[type_key]);
+    } catch (const std::exception& e) {
+      LOG_GENERAL(WARNING, "invalid type: " << type_key << endl << e.what());
+    }
```

Comment:
Should we set `foundVal` to false and also return false here?

Reply:
Same as above.